src/liballoc/slice.rs

```diff
@@ -1328,12 +1334,72 @@ impl<T> [T] {
     /// ```
     #[stable(feature = "slice_sort_by_key", since = "1.7.0")]
     #[inline]
-    pub fn sort_by_key<B, F>(&mut self, mut f: F)
-        where F: FnMut(&T) -> B, B: Ord
+    pub fn sort_by_key<K, F>(&mut self, mut f: F)
+        where F: FnMut(&T) -> K, K: Ord
     {
         merge_sort(self, |a, b| f(a).lt(&f(b)));
     }
 
+    /// Sorts the slice with a key extraction function.
+    ///
+    /// During sorting, the key function is called only once per element.
+    ///
+    /// This sort is stable (i.e. does not reorder equal elements) and `O(m n + n log n)`
+    /// worst-case, where the key function is `O(m)`.
+    ///
+    /// For simple key functions (e.g. functions that are property accesses or
+    /// basic operations), [`sort_by_key`](#method.sort_by_key) is likely to be
+    /// faster.
+    ///
+    /// # Current implementation
+    ///
+    /// The current algorithm is an adaptive, iterative merge sort inspired by
+    /// [timsort](https://en.wikipedia.org/wiki/Timsort).
+    /// It is designed to be very fast in cases where the slice is nearly sorted, or consists of
+    /// two or more sorted sequences concatenated one after another.
+    ///
+    /// The algorithm allocates temporary storage in a `Vec<(K, usize)>` the length of the slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut v = [-5i32, 4, 1, -3, 2];
+    ///
+    /// v.sort_by_cached_key(|k| k.abs());
+    /// assert!(v == [1, 2, -3, 4, -5]);
+    /// ```
+    #[unstable(feature = "slice_sort_by_cached_key", issue = "34447")]
+    #[inline]
+    pub fn sort_by_cached_key<K, F>(&mut self, f: F)
+        where F: FnMut(&T) -> K, K: Ord
+    {
+        // Helper macro for indexing our vector by the smallest possible type, to reduce allocation.
+        macro_rules! sort_by_key {
+            ($t:ty, $slice:ident, $f:ident) => ({
+                let mut indices: Vec<_> =
```

Comment:
Would be nice to be able to use [`SmallVec`](https://github.com/rust-lang/rust/blob/59ccba995de202fb9d9a8d795d2770fb2d199db0/src/librustc_data_structures/small_vec.rs#L30) here and save an allocation when sorting short slices.

Reply:
That's a good idea; I might leave that for a future PR though, yeah.

Reply:
Rather than pulling in a full crate (which is mostly about encapsulating stuff in a single type that can be moved as a unit), you can reproduce that functionality with a couple of stack variables. Something like:

```rust
pub fn foo<T>(s: &[T], f: &Fn(&T) -> T) {
    let iter = s.iter().map(f).enumerate().map(|(i, k)| (k, i as u16));
    let mut vec: Vec<_>;
    let mut array: [_; ARRAY_LEN];
    const ARRAY_LEN: usize = 32;
    let len = iter.len();
    let mapped = if len <= ARRAY_LEN {
        array = unsafe {
            std::mem::uninitialized()
        };
        for (x, slot) in iter.zip(&mut array) {
            *slot = x
        }
        &mut array[..len]
    } else {
        vec = iter.collect();
        &mut vec[..]
    };
    // …
}

Reply:
(Just a caveat, the array code needs to take more precautions to be safe.)

Reply:
@bluss Oh, do you mean panic-safe? Yes, I forgot about that, the array should be inside a `ManuallyDrop`.

Reply:
And use `ptr::write` or the equivalent with a ManuallyDrop around each element.